File: apps/ios/Core/LearningEngine.swift
import Foundation

public struct WordState: Identifiable, Sendable {
    public var id: String
    public var lemma: String
    public var meaning: String
    public var form: String
    public var example: String
    public var bars: Int
    public var understandingCount: Int
    public var independentCount: Int
    public var lastSeen: Date
    public var dueAt: Date
    public var label: String { ["New", "Fragile", "Growing", "Steady"][min(3, max(0, bars))] }
    public var explanation: String {
        if independentCount == 0 { return "Heard or used with support. Try using it in your own words." }
        if bars == 1 { return "Used independently. We’ll bring it back soon." }
        if bars == 2 { return "Recalled on different days. Still worth revisiting." }
        return "Recalled across days and contexts. Strength can fade with time."
    }
}

public struct LearnerState: Sendable {
    public var challenge: Int
    public var observationCount: Int
    public var nextGoal: String
    public var capabilities: [String]
    public var words: [WordState]
    public var levelLabel: String { observationCount < 4 ? "Getting to know you" : "Finding your pace" }
}

public enum LearningEngine {
    public static func validate(_ proposal: Assessment, session: SessionRecord) -> Assessment? {
        guard LanguageRegistry.module(for: session.languageID) != nil,
              let passage = session.passages.first(where: { $0.id == proposal.passageID && $0.speaker == .user }),
              passage.revisionKey == proposal.revisionKey,
              (0...5).contains(proposal.suggestedLevel), proposal.words.count <= 12 else { return nil }
        let allowed = Set(passage.fragments.map(\.id))
        var validated = proposal
        validated.nextGoal = String(validated.nextGoal.prefix(300))
        validated.capability = String(validated.capability.prefix(160))
        validated.words = proposal.words.compactMap { word in
            guard word.language == session.languageID,
                  !word.sourceIDs.isEmpty, Set(word.sourceIDs).isSubset(of: allowed),
                  word.confidence.isFinite, word.confidence >= 0.8, word.confidence <= 1,
                  !word.lemma.isEmpty, word.lemma.count < 100, !word.meaning.isEmpty, word.meaning.count < 180,
                  !word.form.isEmpty, !word.quote.isEmpty,
                  passage.text.localizedCaseInsensitiveContains(word.quote),
                  word.quote.localizedCaseInsensitiveContains(word.form) else { return nil }
            let refs = Passage.join(passage.fragments.filter { word.sourceIDs.contains($0.id) }.map(\.text))
            guard refs.localizedCaseInsensitiveContains(word.quote) else { return nil }
            var result = word
            if result.kind == .independent {
                // A visible meaning or immediate imitation is supporting evidence, never independent recall.
                let recentlyModeled = session.passages.contains {
                    $0.speaker == .assistant && $0.startMS <= passage.startMS && passage.startMS - $0.endMS < 90_000 &&
                    $0.text.localizedCaseInsensitiveContains(word.form)
                }
                if passage.fragments.contains(where: { $0.meaningVisible || $0.typed }) || recentlyModeled { result.kind = .assisted }
            }
            return result
        }
        return validated
    }

    public static func project(_ sessions: [SessionRecord], languageID: String = LanguageRegistry.defaultID, hiddenWords: [String] = [], now: Date = .now) -> LearnerState {
        var level = 0, count = 0, successes = 0
        var nextGoal = "Start with a greeting and one small question. Adjust from what the learner actually says."
        var capabilityEvidence: [String: Set<String>] = [:]
        var events: [String: [(WordProposal, Date, String)]] = [:]
        let calendar = Calendar(identifier: .gregorian)
        for session in sessions.filter({ $0.languageID == languageID }).sorted(by: { $0.startedAt < $1.startedAt }) {
            var seen = Set<String>()
            for raw in session.assessments.sorted(by: { $0.createdAt < $1.createdAt }) {
                guard !seen.contains(raw.passageID), let a = validate(raw, session: session) else { continue }
                seen.insert(raw.passageID)
                count += 1
                if a.outcome == .breakdown { level = max(0, level - 1); successes = 0 }
                else if a.outcome == .success {
                    successes += 1
                    if successes >= 2 { level = min(5, max(level, min(level + 1, a.suggestedLevel))); successes = 0 }
                } else { successes = 0 }
                if !a.nextGoal.isEmpty { nextGoal = a.nextGoal }
                if a.outcome == .success && !a.capability.isEmpty {
                    capabilityEvidence[a.capability, default: []].insert("\(calendar.startOfDay(for: a.createdAt))|\(a.context)")
                }
                var seenWords = Set<String>()
                for word in a.words where !hiddenWords.contains(word.key) && seenWords.insert(word.key).inserted {
                    events[word.key, default: []].append((word, a.createdAt, a.context))
                }
            }
        }
        let words: [WordState] = events.compactMap { key, observations in
            guard let last = observations.last else { return nil }
            let independent = observations.filter { $0.0.kind == .independent }
            let days = Set(independent.map { calendar.startOfDay(for: $0.1) }).count
            let contexts = Set(independent.map { $0.2 }).count
            let lastRecall = independent.last?.1
            var bars = independent.isEmpty ? 0 : 1
            if days >= 2 { bars = 2 }
            if days >= 3 && contexts >= 2 && (independent.last!.1.timeIntervalSince(independent.first!.1) >= 7 * 86400) { bars = 3 }
            let interval = [1.0, 1, 4, 14][bars] * 86400
            let due = (lastRecall ?? last.1).addingTimeInterval(interval)
            if now > due && bars > 1 { bars -= 1 }
            if let lapse = observations.last(where: { $0.0.kind == .lapse }), lapse.1 > (lastRecall ?? .distantPast) { bars = min(bars, 1) }
            return WordState(id: key, lemma: last.0.lemma, meaning: last.0.meaning, form: last.0.form,
                             example: last.0.quote, bars: bars, understandingCount: observations.filter { $0.0.kind == .understanding }.count,
                             independentCount: independent.count, lastSeen: last.1, dueAt: due)
        }.sorted { $0.lastSeen > $1.lastSeen }
        return LearnerState(challenge: level, observationCount: count, nextGoal: nextGoal,
                            capabilities: capabilityEvidence.filter { $0.value.count >= 3 }.keys.sorted(), words: words)
    }
}

File: apps/android/app/src/main/java/chat/mural/core/LearningEngine.kt
package chat.mural.core

data class WordState(
    val id:String,val lemma:String,val meaning:String,val form:String,val example:String,val bars:Int,
    val understandingCount:Int,val independentCount:Int,val lastSeen:Double,val dueAt:Double
) {
    val label get() = listOf("New","Fragile","Growing","Steady")[bars.coerceIn(0,3)]
    val explanation get() = when { independentCount==0 -> "Heard or used with support. Try using it in your own words."
        bars==1 -> "Used independently. We’ll bring it back soon."
        bars==2 -> "Recalled on different days. Still worth revisiting."
        else -> "Recalled across days and contexts. Strength can fade with time." }
}
data class LearnerState(val challenge:Int,val observationCount:Int,val nextGoal:String,val capabilities:List<String>,val words:List<WordState>) {
    val levelLabel get() = if (observationCount < 4) "Getting to know you" else "Finding your pace"
}

object LearningEngine {
    fun validate(proposal:Assessment, session:SessionRecord):Assessment? {
        if (LanguageRegistry.get(session.languageID)==null) return null
        val passage=session.passages.firstOrNull { it.id==proposal.passageID && it.speaker==Speaker.user } ?: return null
        if (passage.revisionKey != proposal.revisionKey || proposal.suggestedLevel !in 0..5 || proposal.words.size>12) return null
        val allowed=passage.fragments.map { it.id }.toSet()
        val join: (List<String>) -> String = if (proposal.textAssemblyVersion == null) Passage::legacyJoin else Passage::join
        val evidenceText = join(passage.fragments.map { it.text })
        val words=proposal.words.mapNotNull { word ->
            if (word.language != session.languageID || word.sourceIDs.isEmpty() || !allowed.containsAll(word.sourceIDs) ||
                !word.confidence.isFinite() || word.confidence !in 0.8..1.0 || word.lemma.isEmpty() || word.lemma.length>=100 ||
                word.meaning.isEmpty() || word.meaning.length>=180 || word.form.isEmpty() || word.quote.isEmpty() ||
                !evidenceText.containsCanonical(word.quote) || !word.quote.containsCanonical(word.form)) return@mapNotNull null
            val refs=join(passage.fragments.filter { word.sourceIDs.contains(it.id) }.map { it.text })
            if (!refs.containsCanonical(word.quote)) return@mapNotNull null
            var out=word
            if (out.kind==EvidenceKind.independent) {
                val modeled=session.passages.any { p -> p.speaker==Speaker.assistant && p.startMS<=passage.startMS && passage.startMS-p.endMS<90000 && join(p.fragments.map { it.text }).containsCanonical(word.form) }
                if (passage.fragments.any { it.meaningVisible || it.typed } || modeled) out=out.copy(kind=EvidenceKind.assisted)
            }
            out
        }
        return proposal.copy(nextGoal=proposal.nextGoal.take(300),capability=proposal.capability.take(160),words=words)
    }
    fun project(sessions:List<SessionRecord>,languageID:String=LanguageRegistry.defaultID,hiddenWords:List<String> = emptyList(),now:Double=nowSeconds()):LearnerState {
        val hidden=hiddenWords.map { it.canonical() }
        var level=0; var count=0; var successes=0
        var nextGoal="Start with a greeting and one small question. Adjust from what the learner actually says."
        val caps=mutableMapOf<String,MutableSet<String>>()
        val events=mutableMapOf<String,MutableList<Triple<WordProposal,Double,String>>>()
        for (session in sessions.filter { it.languageID==languageID }.sortedBy { it.startedAt }) {
            val seen=mutableSetOf<String>()
            for (raw in session.assessments.sortedBy { it.createdAt }) {
                if (raw.passageID in seen) continue
                val a=validate(raw,session) ?: continue
                seen.add(raw.passageID)
                count++
                when(a.outcome) {
                    Outcome.breakdown -> { level=(level-1).coerceAtLeast(0); successes=0 }
                    Outcome.success -> { successes++; if(successes>=2) { level=minOf(5, maxOf(level, minOf(level+1, a.suggestedLevel))); successes=0 } }
                    else -> successes=0
                }
                if(a.nextGoal.isNotEmpty()) nextGoal=a.nextGoal
                if(a.outcome==Outcome.success && a.capability.isNotEmpty()) caps.getOrPut(a.capability){mutableSetOf()}.add("${dayKey(a.createdAt)}|${a.context}")
                val seenWords=mutableSetOf<String>()
                for(word in a.words) if(hidden.none { it==word.key } && seenWords.add(word.key)) events.getOrPut(word.key){mutableListOf()}.add(Triple(word,a.createdAt,a.context))
            }
        }
        val words=events.mapNotNull { (key,obs) ->
            val last=obs.lastOrNull() ?: return@mapNotNull null
            val independent=obs.filter { it.first.kind==EvidenceKind.independent }
            val days=independent.map { dayKey(it.second) }.toSet().size
            val contexts=independent.map { it.third }.toSet().size
            val lastRecall=independent.lastOrNull()?.second
            var bars=if(independent.isEmpty()) 0 else 1
            if(days>=2) bars=2
            if(days>=3 && contexts>=2 && independent.last().second-independent.first().second>=7*86400) bars=3
            val due=(lastRecall ?: last.second)+listOf(1.0,1.0,4.0,14.0)[bars]*86400
            if(now>due && bars>1) bars--
            val lapse=obs.lastOrNull { it.first.kind==EvidenceKind.lapse }
            if(lapse!=null && lapse.second>(lastRecall ?: Double.NEGATIVE_INFINITY)) bars=minOf(bars,1)
            WordState(key,last.first.lemma,last.first.meaning,last.first.form,last.first.quote,bars,
                obs.count { it.first.kind==EvidenceKind.understanding },independent.size,last.second,due)
        }.sortedByDescending { it.lastSeen }
        return LearnerState(level,count,nextGoal,caps.filterValues { it.size>=3 }.keys.sorted(),words)
    }
    private fun dayKey(seconds:Double):String = java.time.Instant.ofEpochSecond((seconds + APPLE_EPOCH_UNIX_SECONDS).toLong()).atZone(java.time.ZoneId.systemDefault()).toLocalDate().toString()
}
